### PR TITLE
[FSDP] Change _create_chunk_dtensor in fsdp/_shard_utils.py to use public API from DTensor

### PR DESCRIPTION
**Pull request by wz337**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #110925
* #110846
* __->__ #110831

This PR:
1) updates _create_chunk_dtensor() in _shard_utils.py to use public APIs from DTensor. This will avoid the global_size calculation error from using DTensor.from_local() for uneven-sharded parameters, as described in https://github.com/pytorch/pytorch/issues/110762 
2) updates test/distributed/fsdp/test_fsdp_dtensor_state_dict.py to include unit test for a model with uneven sharding.

cc. @wanchaol, @fegin 

cc @zhaojuanmao @mrshenli @rohan-varma @awgu @fegin @penguinwu

### CODE REVIEW

File: torch/distributed/fsdp/_shard_utils.py [hunk 1]
```diff
@@ -177,21 +177,17 @@ def _create_chunk_dtensor(
...
-    # Each chunk is a view of the input tensor. If the original tensor change, the view will also be changed.
     # We need to explicitly call .detach() to return a new tensor detached from the current graph.
-    local_tensor = tensor_list[rank].clone().detach()
+    tensor = tensor.clone().detach()
 
     # FSDP placements: [Shard(0)]
     # HSDP placements: [Replicate(), Shard(0)]
-    placements = [Replicate() for _ in range(device_mesh.ndim)]
-    placements[-1] = shard_placement  # type: ignore[call-overload]
-    return DTensor.from_local(local_tensor, device_mesh, placements)
```
Comment (fduwjj) [inline torch/distributed/fsdp/_shard_utils.py:197]: Do we change the logic of local_tensor here? For the old code, it indicates a local shareded Dtensor. But in the new code, we assume it's always local replicate DTensor?

Comment (wz337) [reply torch/distributed/fsdp/_shard_utils.py:197]: > Do we change the logic of local_tensor here? For the old code, it indicates a local shareded Dtensor. But in the new code, we assume it's always local replicate DTensor?

The input tensor for this function here is actually the global tensor we want to shard. We were sharding it in line 180 - 187 in the previous code path(right above) and we were using `shard_placement._split_tensor()`, which is an internal API inside placement_types.py and we don't want to use it. 

Therefore, instead of doing from_local(local_sharded_tensor, device_mesh, Shard(0)) after splitting the tensor, we are using the public DTensor API to do from_local(global_tensor, device_mesh, Replicate()).redistribute(Shard(0)) to create the DTensor.

Comment (fduwjj) [reply torch/distributed/fsdp/_shard_utils.py:197]: I see.. If this is the case, why don't we use the api `distribute_tensor`?

https://github.com/pytorch/pytorch/blob/ddb0c265110cfd053009dee34cbc5151bed8cddc/torch/distributed/_tensor/api.py#L424

File: torch/distributed/fsdp/_shard_utils.py [hunk 2]
```diff
@@ -177,21 +177,17 @@ def _create_chunk_dtensor(
...
-        tensor,
-        device_mesh.size(dim=inner_dim),
-        with_padding=False,
-        contiguous=True,
-    )
-    # We need to explicitly call .clone() here as tensor.chunks() splits a tensor into the specified number of chunks.
-    # Each chunk is a view of the input tensor. If the original tensor change, the view will also be changed.
     # We need to explicitly call .detach() to return a new tensor detached from the current graph.
-    local_tensor = tensor_list[rank].clone().detach()
+    tensor = tensor.clone().detach()
```
Comment (wanchaol) [inline torch/distributed/fsdp/_shard_utils.py:181]: wondering do we still need this? I think we can delete this line as state_dict won't track autograd

File: torch/distributed/fsdp/_shard_utils.py [hunk 3]
```diff
@@ -177,21 +177,17 @@ def _create_chunk_dtensor(
...
 
     # FSDP placements: [Shard(0)]
     # HSDP placements: [Replicate(), Shard(0)]
-    placements = [Replicate() for _ in range(device_mesh.ndim)]
-    placements[-1] = shard_placement  # type: ignore[call-overload]
-    return DTensor.from_local(local_tensor, device_mesh, placements)
+    replicate_placements = [Replicate() for _ in range(device_mesh.ndim)]
```
Comment (wanchaol) [inline torch/distributed/fsdp/_shard_utils.py:185]: do we have a way to detect the current sharding strategy? i.e. if it's fully shard -> [Shard(0)], if it's HSDP, [Replicate(), Shard(0)]

File: torch/distributed/fsdp/_shard_utils.py [hunk 4]
```diff
@@ -177,21 +177,17 @@ def _create_chunk_dtensor(
...
+    shard_placements = [Replicate() for _ in range(device_mesh.ndim)]
+    shard_placements[-1] = DShard(0)  # type: ignore[call-overload]
+    shard_placements = tuple(shard_placements)
```
Comment (wanchaol) [inline torch/distributed/fsdp/_shard_utils.py:188]: I think we don't need this tuple call

File: test/distributed/fsdp/test_fsdp_dtensor_state_dict.py [hunk 5]
```diff
@@ -126,12 +150,17 @@ def test_dtensor_sharded_tensor_state_dict_identical(self, offload_to_cpu):
...
-            self.assertEqual(type(v1), DTensor)
-            self.assertEqual(type(v2), ShardedTensor)
-            # check whether local_tensor are the same
-            self.assertEqual(v1.to_local(), v2.local_tensor())
-            # check whether device are the same
-            self.assertEqual(v1.to_local().device, v2.local_tensor().device)
+            # if the ShardedTensor is an empty shard,
+            # then the local tensor of DTensor should be local_tensor=tensor([])
+            if len(v2.local_shards()) == 0:
+                self.assertEqual(v1.to_local().numel(), 0)
```
Comment (wanchaol) [inline test/distributed/fsdp/test_fsdp_dtensor_state_dict.py:156]: why we are comparing the DTensor path with sharded tensor path? Can be done later but I think we should probably compare with the full state dict path, otherwise the sharded tensor path might become stickier when we get rid of it?

**Review by wanchaol — APPROVED** [review]
lgtm, a few suggestions inlined, thanks for fixing

File: torch/distributed/fsdp/_shard_utils.py (hunk 1 above)
Comment (wanchaol) [reply torch/distributed/fsdp/_shard_utils.py:197]: distribute_tensor does not have assumption about layout current tensor passed in, so for replicate it need to broadcast, sharding it need to scatter, we don't want to call collective here. Also since we already know the tensor is replicated, `from_local` marking it as replicate make more sense.

File: torch/distributed/fsdp/_shard_utils.py (hunk 4 above)
Comment (wz337) [reply torch/distributed/fsdp/_shard_utils.py:188]: Removed in the next PR on top of this.

File: torch/distributed/fsdp/_shard_utils.py (hunk 1 above)
Comment (fduwjj) [reply torch/distributed/fsdp/_shard_utils.py:197]: @wanchaol Well, I mean this looks like a common DTensor operation, would it be better that we can abstract the common logic somewhere in DTensor or we believe this will only be used here for FSDP?

File: torch/distributed/fsdp/_shard_utils.py (hunk 2 above)
Comment (fegin) [reply torch/distributed/fsdp/_shard_utils.py:181]: `clone()` is needed but `detach()` may not be necessary.  However, it doesn't harm to keep `detach()` for explicit show grad is not needed..

File: test/distributed/fsdp/test_fsdp_dtensor_state_dict.py [hunk 6]
```diff
@@ -46,9 +46,29 @@ def get_input(self):
...
 
 
+class TestDummyModelUneven(torch.nn.Module):
+    def __init__(self):
+        super().__init__()
+        torch.manual_seed(0)
+        self.net1 = nn.Sequential(nn.Linear(5, 10), nn.ReLU())
+        self.net2 = nn.Sequential(nn.Linear(10, 15), nn.ReLU())
+        self.net3 = nn.Linear(15, 30)
+        self.net4 = nn.Sequential(nn.ReLU(), nn.Linear(30, 5))
```
Comment (fegin) [inline test/distributed/fsdp/test_fsdp_dtensor_state_dict.py:56]: My original test is pretty funny as there are no Relu for net3. I could not think a reason but a typo. We may want to change it in the future to avoid confusion.

**Review by fegin — APPROVED** [review]

File: torch/distributed/fsdp/_shard_utils.py (hunk 3 above)
Comment (wz337) [reply torch/distributed/fsdp/_shard_utils.py:185]: > do we have a way to detect the current sharding strategy? i.e. if it's fully shard -> [Shard(0)], if it's HSDP, [Replicate(), Shard(0)]

Agree that this is more explicit and easier to read. We can update all the use case for FSDP and HSDP in later PR, but for 2D FSDP + TP, we will still need to do something like this.

**Comment by malfet** [comment]
Closing as merged as  https://github.com/pytorch/pytorch/commit/d9eb5a57aade6a3828cbf151b721c060bb8848df by https://github.com/pytorch/pytorch/actions/runs/6474766763

File: torch/distributed/fsdp/_shard_utils.py (hunk 2 above)
Comment (lazychao) [reply torch/distributed/fsdp/_shard_utils.py:181]: Hi, why clone() is needed here？ DTensor.redistribute will also do clone() in _replicate_to_shard
